src/compiler/checker.ts

```diff
@@ -26900,6 +26900,20 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         return propType && getConstituentTypeForKeyType(unionType, propType);
     }
 
+    function getMatchingUnionConstituentForArrayLiteral(unionType: UnionType, node: ArrayLiteralExpression) {
+        const resolvedElements = node.elements.map(el => getContextFreeTypeOfExpression(el));
+        for (const type of unionType.types) {
+            if (!isTupleType(type)) continue;
+            const unionElements = getElementTypes(type);
+            if (unionElements.length !== resolvedElements.length) continue;

```

Comment:
This is not quite right. You need to experiment more with this and try to create multiple test cases mixing tuples with different fixed sizes and also the ones with variadic sizes. It would probably be the easiest to only reason about the *leading* fixed elements.

Reply:
https://github.com/microsoft/TypeScript/pull/58571/commits/79a4a79d1b3cb2567642d7e5bc40e276c1bfc23a

i didn't see an obvious way to determine the number of leading elements so i removed that early exit